File: pytet_v0.5/ctetris.py
from tetris import *
from matrix import *

class CTetris(Tetris):
	setOfCBlockObjects = 0

	@classmethod
	def init(cls, setOfBlockArrays): 
		Tetris.init(setOfBlockArrays)
		CTetris.setOfCBlockObjects = [[0]* Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]

		for i in range(Tetris.nBlockTypes):
			for j in range(Tetris.nBlockDegrees):
				obj = Matrix(setOfBlockArrays[i][j])
				obj.mulc(i+1) #블록별로 색상을 다르게 하기 위함
				CTetris.setOfCBlockObjects[i][j] = obj 
				#블록별로 색상 다르게 하여 CTetris 클래스 초기설정(블록 세트 설정)
		return
	
	def __init__(self, cy, cx): #CTetris 생성자
		Tetris.__init__(self, cy, cx) #Tetris상속받아 객체 생성
		arrayScreen = self.createArrayScreen() #CTetris의 arrayScreen=Tetris 상속받아 만든 ArrayScreen
		self.iCScreen = Matrix(arrayScreen) #arrayScreen을 Matrix type으로 받아 iCScreen설정
		self.oCScreen = Matrix(self.iCScreen) #iCScreen을 Matrix type으로 받아 oCScreen설정
		return

	def accept(self, key): #
		if key >= '0' and key <= '6':
			if self.justStarted == False:
				self.deleteFullLines()
			self.iCScreen = Matrix(self.oCScreen)

		state = Tetris.accept(self, key)
		
		currCBlk = CTetris.setOfCBlockObjects[self.idxBlockType][self.idxBlockDegree]
		tempBlk = self.iCScreen.clip(self.top, self.left, 
									self.top + currCBlk.get_dy(), 
									self.left + currCBlk.get_dx())
		tempBlk = tempBlk + currCBlk

		self.oCScreen = Matrix(self.iCScreen)
		self.oCScreen.paste(tempBlk, self.top, self.left)
		return state
		
	def deleteFullLines(self):
		nDeleted = 0
		nScanned = self.currBlk.get_dy()

		if self.top + self.currBlk.get_dy() - 1 >= self.iScreenDy:
			nScanned -= (self.top + self.currBlk.get_dy() - self.iScreenDy)

		zero = Matrix([[ 0 for x in range(0, (self.iScreenDx - 2*Tetris.iScreenDw))]])
		for y in range(nScanned - 1, -1, -1):
			cy = self.top + y + nDeleted
			line = self.oScreen.clip(cy, 0, cy+1, self.oScreen.get_dx())
			if line.sum() == self.oScreen.get_dx():
				### Tetris screen
				temp = self.oScreen.clip(0, 0, cy, self.oScreen.get_dx())
				self.oScreen.paste(temp, 1, 0)
				self.oScreen.paste(zero, 0, Tetris.iScreenDw)

				### CTetris screen
				temp = self.oCScreen.clip(0, 0, cy, self.oCScreen.get_dx())
				self.oCScreen.paste(temp, 1, 0)
				self.oCScreen.paste(zero, 0, Tetris.iScreenDw)

				nDeleted += 1
		return

### end of class CTetris():

File: mobilePro/HW1/CTetris.java
public class CTetris extends Tetris{
    private static Matrix[][] setOfCBlockObjects;
    private Matrix iCScreen;
    public Matrix oCScreen;
    private Matrix currCBlk;

    private static void changeToTetris(int[][][][] setOfArrays) throws Exception {
        int ntypes = setOfArrays.length;
        int ndegrees = setOfArrays[0].length;
        for (int t = 0; t < ntypes; t++) {
            for (int d = 0; d < ndegrees; d++) {
                for(int i=0; i<setOfArrays[t][d].length; i++){
                    for(int j=0; j<setOfArrays[t][d].length; j++){
                        if(setOfArrays[t][d][i][j]!=0)
                            setOfArrays[t][d][i][j]=1;
                        else
                            setOfArrays[t][d][i][j]=0;
                    }
                }
            }
        }
    }

    public static void init(int[][][][] setOfBlockArrays) throws Exception {
        nBlockTypes = setOfBlockArrays.length;
        nBlockDegrees = setOfBlockArrays[0].length;
        setOfCBlockObjects=createSetOfBlocks(setOfBlockArrays);
        changeToTetris(setOfBlockArrays);
        Tetris.init(setOfBlockArrays);
    }

    public CTetris(int cy, int cx) throws Exception {
        super(cy, cx);
        iCScreen = new Matrix(iScreen);
        oCScreen = new Matrix(oScreen);
    }

    public TetrisState accept(char key) throws Exception {
        Matrix tempCBlk;
        if (state == TetrisState.NewBlock) {
            oCScreen=deleteFullLines(oCScreen, currCBlk, top, iScreenDy, iScreenDx, iScreenDw);
            iCScreen = new Matrix(oCScreen);
        }
        state = super.accept(key);
        currCBlk = setOfCBlockObjects[idxBlockType][idxBlockDegree];
        tempCBlk = iCScreen.clip(top, left, top + currCBlk.get_dy(), left + currCBlk.get_dx());
        tempCBlk = tempCBlk.add(currCBlk);
        oCScreen = new Matrix(iCScreen);
        oCScreen.paste(tempCBlk, top, left); //System.out.println();
        return state;
    }
    private Matrix deleteFullLines(Matrix screen, Matrix blk, int top, int dy, int dx, int dw) throws Exception {
        Matrix line, zero, temp, ctmp;
        if (blk == null) return screen; // called right after the game starts!!
        int cy, y, nDeleted = 0,nScanned = blk.get_dy();
        if (top + blk.get_dy() - 1 >= dy)
            nScanned -= (top + blk.get_dy() - dy);
        zero = new Matrix(1, dx - 2*dw);
        for (y = nScanned - 1; y >= 0 ; y--) {
            cy = top + y + nDeleted;
            line = oScreen.clip(cy, 0, cy + 1, screen.get_dx());
            if (line.sum() == oScreen.get_dx()) {
                temp = oScreen.clip(0, 0, cy, screen.get_dx());
                ctmp=screen.clip(0, 0, cy, screen.get_dx());
                oScreen.paste(temp, 1, 0);
                screen.paste(ctmp, 1, 0);
                oScreen.paste(zero, 0, dw);
                screen.paste(zero, 0, dw);

                nDeleted++;
            }
        }
        return screen;
    }
}